site/lib/src/components/common/client/api_link_tooltip.dart

```diff
@@ -0,0 +1,252 @@
+
+    if (extractedHeader == null && extractedDescription == null) {
+      return;
+    }
+
+    if (!mounted) return;
+    setState(() {
+      tooltipContent = fragment([
+        if (extractedHeader != null)
+          span(classes: 'tooltip-header', [raw(extractedHeader)]),
+        if (extractedDescription != null)
+          span(classes: 'tooltip-content', [raw(extractedDescription)]),
+      ]);
+
+      isTouchscreen = web.window.matchMedia('(pointer: coarse)').matches;
+    });
+
+    context.binding.addPostFrameCallback(ensureVisible);
+
+    // Reposition tooltips on window resize.
+    web.EventStreamProviders.resizeEvent.forTarget(web.window).listen((_) {
+      ensureVisible();
+    });
+  }
+
+  /// Adjust the tooltip position to ensure it is fully inside the
+  /// ancestor .content element.
+  void ensureVisible() {
+    final target = targetKey.currentNode;
+    final tooltip = tooltipKey.currentNode;
+    if (tooltip == null || target == null) return;
+
+    setState(() {
+      tooltipOffset = calculateTooltipOffset(target, tooltip);
+    });
+  }
+
+  @override
+  Component build(BuildContext context) {
+    Component? tooltip;
+
+    if (tooltipContent != null) {
+      tooltip = span(
+        key: tooltipKey,
+        classes: ['tooltip', 'large', if (isVisible) 'visible'].toClasses,
+        styles: Styles(
+          raw: {
+            'left': tooltipOffset == 0
+                ? '50%'
+                : 'calc(50% ${tooltipOffset > 0 ? '+' : '-'} ${tooltipOffset.abs()}px)',
+          },
+        ),
+        [tooltipContent!],
+      );
+
+      if (isTouchscreen) {
+        tooltip = GlobalEventListener(
+          // Close tooltip when clicking outside of this wrapper.
+          onClick: (e) {
+            if (wrapperKey.currentNode?.contains(e.target as web.Node?) ==
+                true) {
+              return;
+            }
+            setState(() => isVisible = false);
+          },
+          // On touchscreen devices, close tooltips when scrolling.
+          onScroll: (_) {
+            setState(() => isVisible = false);
+          },
+          tooltip,
+        );
+      }
+    }
+
+    return span(key: wrapperKey, classes: 'tooltip-wrapper', [
+      a(
+        key: targetKey,
+        href: component.url,
+        classes: 'tooltip-target',
+        events: {
+          if (tooltipContent != null && isTouchscreen)
+            'click': (event) {
+              if (!isVisible) {
+                setState(() => isVisible = true);
+                event.preventDefault();
+              }
+            },
+        },
+        [
+          code([text(component.text)]),
+        ],
+      ),
+      ?tooltip,
+    ]);
+  }
+}
+
+const contentId = 'dartdoc-main-content';
+// This seems to be a good limit to avoid overly small or large tooltips.
+const maxDescriptionLength = 400;
+const minTrailingParagraphLength = 20;
+
+@awaitNotRequired
+Future<(String?, String?)> scrapeApiDocs(String url) async {
```

Comment:
Why scrape the docs? We also control the generation of the docs, could this snippet be added to the API doc generation? This sounds like useful functionality for the whole ecosystem, not just api.fluter.dev

Reply:
> We also control the generation of the docs, could this snippet be added to the API doc generation?

Good point! I suppose it could or at least more easily extracted from a general Markdown version of each page.

My hopeful thinking was to extract the API information directly with https://github.com/dart-lang/labs/tree/main/pkgs/jot or something else more we build. 

For now this scraping is a great way to quickly show off what this could look like though :D